tests/components/vizio/test_config_flow.py

```diff
@@ -0,0 +1,302 @@
+"""Tests for Vizio config flow."""
+import logging
+
+from asynctest import patch
+import voluptuous as vol
+
+from homeassistant import data_entry_flow
+from homeassistant.components.vizio import VIZIO_SCHEMA
+from homeassistant.components.vizio.const import (
+    CONF_VOLUME_STEP,
+    DEFAULT_NAME,
+    DEFAULT_VOLUME_STEP,
+    DOMAIN,
+)
+from homeassistant.const import (
+    CONF_ACCESS_TOKEN,
+    CONF_DEVICE_CLASS,
+    CONF_HOST,
+    CONF_NAME,
+)
+from homeassistant.helpers.typing import HomeAssistantType
+
+from tests.common import MockConfigEntry
+
+_LOGGER = logging.getLogger(__name__)
+
+NAME = "Vizio"
+NAME2 = "Vizio2"
+HOST = "192.168.1.1:9000"
+HOST2 = "192.168.1.2:9000"
+DEVICE_CLASS_TV = "tv"
+DEVICE_CLASS_SOUNDBAR = "soundbar"
+ACCESS_TOKEN = "deadbeef"
+VOLUME_STEP = 2
+UNIQUE_ID = "testid"
+UNIQUE_ID2 = "testid2"
+
+MOCK_USER_VALID_TV_ENTRY = {
+    CONF_NAME: NAME2,
+    CONF_HOST: HOST2,
+    CONF_DEVICE_CLASS: DEVICE_CLASS_TV,
+    CONF_ACCESS_TOKEN: ACCESS_TOKEN,
+}
+
+MOCK_IMPORT_VALID_TV_ENTRY = {
+    CONF_NAME: NAME2,
+    CONF_HOST: HOST2,
+    CONF_DEVICE_CLASS: DEVICE_CLASS_TV,
+    CONF_ACCESS_TOKEN: ACCESS_TOKEN,
+    CONF_VOLUME_STEP: VOLUME_STEP,
+}
+
+MOCK_INVALID_TV_ENTRY = {
+    CONF_NAME: NAME,
+    CONF_HOST: HOST,
+    CONF_DEVICE_CLASS: DEVICE_CLASS_TV,
+}
+
+MOCK_SOUNDBAR_ENTRY = {
+    CONF_NAME: NAME,
+    CONF_HOST: HOST,
+    CONF_DEVICE_CLASS: DEVICE_CLASS_SOUNDBAR,
+}
+
+
+class MockInput(object):
+    """Mock Vizio Input."""
+
+    def __init__(self, name):
+        """Initialize Mock Vizio Input."""
+        self.name = name
+
+
+async def test_user_flow(hass: HomeAssistantType) -> None:
+    """Test user config flow."""
+
+    # test form shows
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": "user"}
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    # test valid options
+    with patch("pyvizio.vizio.VizioAsync.can_connect", return_value=True), patch(
+        "pyvizio.vizio.VizioAsync.get_esn", return_value=UNIQUE_ID
+    ), patch("pyvizio.vizio.VizioAsync.get_power_state", return_value=True), patch(
+        "pyvizio.vizio.VizioAsync.get_current_volume", return_value=0
+    ), patch(
+        "pyvizio.vizio.VizioAsync.get_current_input", return_value="HDMI"
+    ), patch(
+        "pyvizio.vizio.VizioAsync.get_inputs", return_value=[MockInput("HDMI")]
```

Comment:
Consider breaking these patches out into a single pytest fixture. Then we can yield the return value of the patched class, ie the mock `VizioAsync` instance. That way we can set different method return values as needed inside the different tests that need to patch this class.

```py
@pytest.fixture(name="vizio_device")
def vizio_device_fixture():
    """Mock VizioAsync device""".
    with patch("homeassistant.components.vizio.config_flow.VizioAsync", autospec=True) as vizio_class:
        yield vizio_class.return_value
```

Reply:
We also may need to patch the vizio integration `async_setup_entry` to avoid setting up the complete integration. We don't need or want that to test the config flow.